File: 589.n-ary-tree-preorder-traversal/solution.kt
/**
 * Definition for a Node.
 * class Node(var `val`: Int) {
 *     var children: List<Node?> = listOf()
 * }
 */

class Solution {
    fun preorder(root: Node?): List<Int> {
        if (root == null) return listOf()

        var stack = mutableListOf(root)
        var results = mutableListOf<Int>()
        while (stack.size > 0) {
            val curr = stack.removeAt(stack.size - 1)
            results.add(curr.`val`)
            for (i in curr.children.size - 1 downTo 0) {
                stack.add(curr.children[i]!!)
            }
        }
        return results
    }
}

File: 589.n-ary-tree-preorder-traversal/solution.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val=None, children=None):
        self.val = val
        self.children = children
"""

class Solution:
    def preorder(self, root: 'Node') -> List[int]:
        if root is None:
            return []

        result = []
        stack = [root]
        while (stack):
            curr = stack.pop()
            result.append(curr.val)
            while(curr.children):
                stack.append(curr.children.pop())

        return result
